pkg/resourcesets/collector.go

```diff
@@ -248,157 +174,147 @@ func (h *ResourceHandler) filterByLabel(ctx context.Context, dr dynamic.Resource
-		for _, name := range filter.ResourceNames {
-			allowedNames[name] = true
+	// Treat no filter.ResourceNameRegexp as "." when filter.ExcludeResourceNameRegexp is specified.
+	// And precompile the patterns before looping through the resources
+	resourceNameRegexpStr := filter.ResourceNameRegexp
+	excludeResourceNameRegexpStr := filter.ExcludeResourceNameRegexp
+	if excludeResourceNameRegexpStr != "" && resourceNameRegexpStr == "" {
+		resourceNameRegexpStr = "."
+	}
+	if resourceNameRegexpStr != "" {
+		resourceNameRegexp, err = regexp.Compile(resourceNameRegexpStr)
+		if err != nil {
+			return filteredByName, err
 		}
-		for _, resObj := range resourceObjectsList {
-			metadata := resObj.Object["metadata"].(map[string]interface{})
-			name := metadata["name"].(string)
-			if allowedNames[name] {
-				filteredByResourceNames = append(filteredByResourceNames, resObj)
-			}
+	}
+	if excludeResourceNameRegexpStr != "" {
+		excludeResourceNameRegexp, err = regexp.Compile(excludeResourceNameRegexpStr)
+		if err != nil {
+			return filteredByName, err
 		}
+	}
 
-		if len(filteredByResourceNames) == 0 {
-			// exact names were provided, but no resources found by that name
-			// so return anything obtained from matching resource names by regex
-			// if that list is empty too, means nothing matched the filters
-			return filteredByName, nil
-		}
-		if len(filteredByNameMap) > 0 {
-			// avoid duplicates
-			for _, resObj := range filteredByResourceNames {
-				key := resObj.Object["filterUid"].(string)
-				if !filteredByNameMap[key] {
-					filteredByName = append(filteredByName, resObj)
-				}
+	logrus.Debugf("Using ResourceNameRegexp [%s] to filter resource names", filter.ResourceNameRegexp)
+
+	for _, resObj := range resourceObjectsList {
+		metadata := resObj.Object["metadata"].(map[string]interface{})
+		resourceName := metadata["name"].(string)
+		// Do the cheaper check first: is this in the list of ResourceNames we want (so ignore exclusion)
+		if allowedNames[resourceName] {
+			filteredByName = append(filteredByName, resObj)
+			continue
+		}
+		if resourceNameRegexpStr != "" {
+			nameMatched := resourceNameRegexpStr == "." || resourceNameRegexp.MatchString(resourceName)
+			if !nameMatched {
+				logrus.Debugf("Skipping [%s] because it did not match ResourceNameRegexp [%s]", resourceName, filter.ResourceNameRegexp)
+			} else if excludeResourceNameRegexp != nil && excludeResourceNameRegexp.MatchString(resourceName) {
+				logrus.Debugf("Skipping [%s] because it matched both ResourceNameRegexp and ExcludeResourceNameRegexp [%s]", resourceName, filter.ExcludeResourceNameRegexp)
+			} else {
+				filteredByName = append(filteredByName, resObj)
 			}
-		} else {
-			filteredByName = filteredByResourceNames
 		}
-
 	}
+
 	return filteredByName, nil
 }
 
 func (h *ResourceHandler) filterByNamespace(filter v1.ResourceSelector, filteredByName []unstructured.Unstructured) ([]unstructured.Unstructured, error) {
```

Comment:
For this method we only concern with:

```
	Namespaces                []string              `json:"namespaces,omitempty"`
	NamespaceRegexp           string                `json:"namespaceRegexp,omitempty"`
```

And have no exclude mechanism. So essentially the logic for this method should just be:

`MatchNamespacesList || MatchNamespaceRegex`

Reply:
True. I mixed up excludeResourceNameRegexp with a non-existent `excludeNamespaceRegexp`